server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1411,14 +1411,16 @@ public long recoverLocallyUpToGlobalCheckpoint() {
             logger.debug("skip local recovery as failed to find the safe commit", e);
             return UNASSIGNED_SEQ_NO;
         }
-        if (safeCommit.isPresent() == false) {
-            logger.trace("skip local recovery as no safe commit found");
-            return UNASSIGNED_SEQ_NO;
-        }
-        assert safeCommit.get().localCheckpoint <= globalCheckpoint : safeCommit.get().localCheckpoint + " > " + globalCheckpoint;
         try {
             maybeCheckIndex(); // check index here and won't do it again if ops-based recovery occurs
             recoveryState.setStage(RecoveryState.Stage.TRANSLOG);
+            if (safeCommit.isPresent() == false) {
+                assert globalCheckpoint == UNASSIGNED_SEQ_NO || indexSettings.getIndexVersionCreated().before(Version.V_6_2_0) :
```

Comment:
The new test found this issue where the index is synced flush, but the global checkpoint is still unassigned.

Reply:
I'm not sure I understand what issue is being addressed here. How is moving this condition further down (after maybeCheckIndex) helping?
Is the issue that we have not properly moved to the translog stage?

Reply:
Also, which part of the new tests show this issue, and is it something that can also be triggered with a single restart?

Reply:
Sorry, I should have explained better. 

> Is the issue that we have not properly moved to the translog stage?

That's correct. Previously, we do not move the recovery stage from INDEX to TRANSLOG if we don't have the safe commit, which can be the case if the index was created before 6.2 or the global checkpoint is still unassigned. Here we expect a file-based recovery to happen, and we will move the recovery stage to TRANSLOG in the [clean files](https://github.com/dnhatn/elasticsearch/blob/df46848fb09b1da3ea350c95f2843f6d1e1ac314/server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java#L435) step. However, if the shard has a synced flush, we won't execute the clean files step and trip the [assertion](https://github.com/elastic/elasticsearch/blob/c60082b3931dcacbb318ce621ad63066b4b58d9a/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L1603).

> Also, which part of the new tests show this issue, and is it something that can also be triggered with a single restart?

Yes, I will add it to the full cluster restart suite.